test/extensions_test.dart

```diff
@@ -989,6 +989,52 @@ void main() {
         expect(iterable(['4', '3', '2', '1']).isSorted(), false);
       });
     });
+    group('.choose', () {
+      test('errors', () {
+        expect(() => iterable([]).choose(1), throwsStateError);
+        expect(() => iterable([1]).choose(-1), throwsRangeError);
+        expect(() => iterable([1]).choose(2), throwsStateError);
+      });
+      test('empty', () {
+        var empty = iterable(<int>[]);
+        expect(empty.choose(0), []);
+      });
+      test('single', () {
+        var single = iterable([1]);
+        expect(single.choose(0), []);
+        expect(single.choose(1), [1]);
+      });
+      test('multiple', () {
+        var multiple = iterable([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
+        expect(multiple.choose(0), []);
+        var one = multiple.choose(1);
+        expect(one.length, 1);
+        expect(one.first, inInclusiveRange(1, 10));
+        var seen = <int>{};
+        do {
+          var some = multiple.choose(3);
+          expect(some.length, 3);
+          expect(some[0], inInclusiveRange(1, 10));
+          expect(some[1], inInclusiveRange(1, 10));
+          expect(some[2], inInclusiveRange(1, 10));
+          expect(some[0], isNot(some[1]));
+          expect(some[0], isNot(some[2]));
+          expect(some[1], isNot(some[2]));
+          seen.addAll(some);
+        } while (seen.length < 10);
+        // Should eventually terminate.
```

Comment:
note that since there is a call to `expect` in the loop, if for some reason this _doesn't_ terminate it also won't time out in the test runner.

I don't think this is a real risk here, but the pattern does make me nervous if we applied it elsewhere.

Reply:
It should probably make us nervous here too, I'll unroll the loop once so there are no expects inside the loop.